mist-core/src/main/java/edu/snu/mist/core/master/SingleNodeRecoveryScheduler.java

```diff
@@ -0,0 +1,148 @@
+
+  private static final Logger LOG = Logger.getLogger(RecoveryScheduler.class.getName());
+
+  /**
+   * The map which contains all the failed groups.
+   */
+  private ConcurrentMap<String, GroupStats> allFailedGroups;
+
+  /**
+   * The map which contains the groups to be recovered.
+   */
+  private ConcurrentMap<String, GroupStats> recoveryGroups;
+
+  /**
+   * The shared taskStatsMap.
+   */
+  private TaskStatsMap taskStatsMap;
+
+  /**
+   * The shared map which contains Avro proxies to task.
+   */
+  private ProxyToTaskMap proxyToTaskMap;
+
+  /**
+   * The proxy to the recovery task.
+   */
+  private MasterToTaskMessage proxyToRecoveryTask;
+
+  /**
+   * The lock which is used for conditional variable.
+   */
+  private final Lock lock;
+
+  /**
+   * The conditional variable which synchronizes the recovery process.
+   */
+  private final Condition notInRecoveryProcess;
+
+  /**
+   * The boolean variable which indicates whether now in recovery process or not.
+   */
+  private boolean isRecovering;
+
+  @Inject
+  private SingleNodeRecoveryScheduler(
+      final TaskStatsMap taskStatsMap,
+      final ProxyToTaskMap proxyToTaskMap) {
+    this.allFailedGroups = new ConcurrentHashMap<>();
+    this.recoveryGroups = null;
+    this.taskStatsMap = taskStatsMap;
+    this.proxyToTaskMap = proxyToTaskMap;
+    this.lock = new ReentrantLock();
+    this.notInRecoveryProcess = this.lock.newCondition();
+    this.isRecovering = false;
+  }
+
+  @Override
+  public synchronized void addFailedGroups(final Map<String, GroupStats> failedGroups) {
+    this.allFailedGroups.putAll(failedGroups);
+  }
+
+  @Override
+  public synchronized void startRecovery() {
+    try {
+      lock.lock();
+      while (isRecovering) {
+        notInRecoveryProcess.await();
+      }
+      isRecovering = true;
+      this.recoveryGroups = allFailedGroups;
+      this.allFailedGroups = new ConcurrentHashMap<>();
+      // Select the newly allocated task with the least load for recovery
+      double minLoad = Double.MAX_VALUE;
+      proxyToRecoveryTask = null;
+      for (final Map.Entry<String, MasterToTaskMessage> entry : proxyToTaskMap.entrySet()) {
+        final String taskHostname = entry.getKey();
+        final double taskLoad = taskStatsMap.get(taskHostname).getTaskLoad();
+        if (taskLoad < minLoad) {
+          minLoad = taskLoad;
+          proxyToRecoveryTask = entry.getValue();
+        }
+      }
+      proxyToRecoveryTask.startRecovery();
+    } catch (final InterruptedException e) {
+      LOG.log(Level.SEVERE, "Recovery has been interrupted. " + e.toString());
+    } catch (final AvroRemoteException e) {
+      LOG.log(Level.SEVERE, "Start recovery through avro server has failed! " + e.toString());
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public List<String> getRecoveringGroups(final String taskHostname) {
+    // No more groups to be recovered! Recovery is done!
+    if (recoveryGroups.isEmpty()) {
```

Comment:
Also, `recoveryGroups` can be concurrently read and modified. We need to use `AtomicReference` or just make `getRecoveringGroups` synchronized.